### PR TITLE
connect: Implemented exercise. (Closes #1743)

### LINKED ISSUE #1743: connect: implement exercise
The exercise **connect** has not been implemented yet for the Java track. 
The description of the exercise can be found in the [problem specification repository](https://github.com/exercism/problem-specifications/tree/master/exercises/connect). 

How to implement a new exercise for the Java track is described in detail in [CONTRIBUTING.md](https://github.com/exercism/java/blob/master/CONTRIBUTING.md#adding-a-new-exercise). Please have a look there first before starting working on the exercise. 
Also please make sure it is clear that you are currently working on this issue, either by asking to be assigned to it, or by opening an empty PR. 

When opening an PR, please reference this issue using any of the [closing keywords](https://help.github.com/en/articles/closing-issues-using-keywords). 

If you have had a look at the exercise description and you concluded that the exercise might not be possible to implement in the Java language, please leave a comment and describe the problem. 

In case you have any further questions, feel free to ask here.

### PR DESCRIPTION
**Pull request by tr808axm**
This PR implements the connect exercise. For the exercise configuration parameters in `config.json`, I used the [values from the JavaScript track](https://github.com/exercism/javascript/blob/master/config.json#L994).

---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/master/POLICIES.md#event-checklist)

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -1441,6 +1441,21 @@
       "difficulty": 0,
       "topics": null,
       "deprecated": true
+    },
+    {
+      "slug": "connect",
+      "uuid": "1a714bd0-ed15-4b19-8f54-19c6e269763b",
+      "core": false,
+      "unlocked_by": "grade-school",
```
Comment (lemoncurry) [inline config.json]: Exercises are sorted by difficulty level. Please put it the the correct difficulty level.

Also, please correct the error "The exercise 'connect' is being unlocked by a non-core exercise. Non-core exercises can only be unlocked by core exercises."

File: exercises/connect/.meta/src/reference/java/Connect.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
+class Connect {
+    String findWinner(String[] board) {
+        // parse
```
Comment (lemoncurry) [inline exercises/connect/.meta/src/reference/java/Connect.java]: I would propose to extract this as method.

File: exercises/connect/.meta/src/reference/java/Connect.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        for (int i = 0; i < columnCount; i++) {
+            if (values[0][i].equals("O")) values[0][i] = "o";
+        }
+        // X starts on the left
+        for (int i = 0; i < lineCount; i++) {
+            if (values[i][0].equals("X")) values[i][0] = "x";
+        }
+
+
+        // go through and connect
```
Comment (lemoncurry) [inline exercises/connect/.meta/src/reference/java/Connect.java]: Here also I would suggest to extract this as method.

File: exercises/connect/.meta/src/reference/java/Connect.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        // right
+        if (!isLastColumn && values[line][column + 1].equals(character)) return true;
+        // bottom left
+        if (!isLastLine && !isFirstColumn && values[line + 1][column - 1].equals(character)) return true;
+        // bottom right
+        if (!isLastLine && values[line + 1][column].equals(character)) return true;
+        // no match
+        return false;
+    }
+}
```
Comment (lemoncurry) [inline exercises/connect/.meta/src/reference/java/Connect.java]: Add new line

File: exercises/connect/src/main/java/Connect.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class Connect {
+    String findWinner(String[] board) {
+        throw new UnsupportedOperationException("Delete this statement and write your own implementation.");
+    }
+}
```
Comment (lemoncurry) [inline exercises/connect/src/main/java/Connect.java]: Add new line and use the standard template for stub files for exercises with difficulty > 4.

File: exercises/connect/src/test/java/ConnectTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                "  . . . . .",
+                "   . . . . .",
+                "    . . . . ."
+        };
+        assertThat(new Connect().findWinner(board), is(""));
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testOnlySmallX() {
```
Comment (lemoncurry) [inline exercises/connect/src/test/java/ConnectTest.java]: Please check that the test method names align with the descriptions provided in the canonical data.

File: exercises/connect/src/test/java/ConnectTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
+import org.junit.Ignore;
```
Comment (lemoncurry) [inline exercises/connect/src/test/java/ConnectTest.java]: Add a version file to .meta

File: exercises/connect/.meta/src/reference/java/Connect.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        // X wins on the right
+        for (int i = 0; i < lineCount; i++) {
+            if (values[i][columnCount - 1].equals("x")) winner = "X";
+        }
+
+        return winner;
+    }
+
+    boolean checkSurroundingsFor(String[][] values, int line, int column, String character) {
```
Comment (lemoncurry) [inline exercises/connect/.meta/src/reference/java/Connect.java]: This method appears to be quite complicated and hard to read. Could you please rethink the implementation here?
For example it helps to extract more methods as I already asked to do in the code above.

In the C# track is a good example of an implementation that uses many small methods and more descriptive method and variable names to improve readability.

**Comment by jmrunkle** [comment]
@tr808axm - are you still working on this PR? If not, we might close it to clean up the backlog.

**Comment by jmrunkle** [comment]
> @tr808axm - are you still working on this PR? If not, we might close it to clean up the backlog.

Please reopen this PR if you are still actively working on it and we can continue the review.

**Pull request closed without merging**